rgw: user accounts implementation

cbodley opened the pull request:
user account design document in https://github.com/ceph/ceph/pull/54045

this revives previous work from https://github.com/ceph/ceph/pull/46373 (which iself was based on Abhishek's https://github.com/ceph/ceph/pull/35726). that became too painful to faithfully rebase, so i've reimplemented parts of it instead; i'll be sure to credit Abhishek for his work in the final commits

# TODO
## Account metadata
- [x] `struct RGWAccountInfo`
- [x] requirements for account ids and name (`rgw::account::validate_id/name()`)
- [x] `rgw::sal::Driver` interfaces to read/write
- [x] admin ops to create/modify/list/remove account metadata
- [x] add email, and ensure it's globally unique (including user emails)

## User metadata
- [x] add `RGWUserInfo::account_id` member
- [x] support --account-id argument for `user modify` to migrate existing legacy/non-account users into an account. account id can't change after
- [x] migrating a user should also migrate the user's buckets, changing their owner to the account
- [x] add other metadata fields needed for IAM User APIs (Path/Arn/CreateDate)
- [x] match user paths in iam policy principals

## Role metadata
- [x] add `RGWRoleInfo::account_id` member
- [x] index account roles in the {account}.roles omap
- [x] within an account, role names don't have to be globally unique - don't store the 'path' object in rados, use omap for listing

## Group metadata
- [x] `struct RGWGroupInfo`
- [x] `rgw::sal::Driver` interfaces to read/write/list
- [x] metadata handler for multisite replication

## Topic metadata
- [x] change owner from rgw_user to rgw_owner
- [x] index account topics in the {account}.topics omap

## RadosStore
- [x] implement account interfaces for `RadosStore`
- [x] use `cls_user` to track stats in `{account}.buckets` the same way users do in `{user}.buckets`
- [x] extend `cls_user` to track `{account}.users`, `{account}.roles`, `{account}.groups` (started in https://github.com/ceph/ceph/pull/54563)
- [x] `RGWBucketMetadataHandler`: on change to `RGWBucketEntryPoint::owner`, update linkage in `{user}.buckets` and/or `{account}.buckets` objects via cls_user
- [x] `RGWUserMetadataHandler`: on change to `RGWUserInfo::account_id`, update linkage in `{account}.users` via cls_user

## ACLs
- [x] ACLOwner and CanonicalUser grants can either be a user id string or an account id string
- [x] AmazonCustomerByEmail supports account emails in addition to user emails
- [x] resources created by a user in an account use account id for default ACLOwner and FULL_CONTROL Grant
- [x] use account name for DisplayName

## IAM Policy
- [x] Principal: disambiguate 'tenant' and 'account' principals using `rgw::account::validate_id()`
- [x] match 'account' principals against `RGWUserInfo::account_id` instead of `RGWUserInfo::user_id.tenant`
- [x] load group policies for use in verify_permission
- [x] support [managed policies](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_managed-vs-inline.html#aws-managed-policies) like AmazonS3FullAccess for AttachUserPolicy/AttachRolePolicy
- [x] implement [cross-account policy evaluation](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies-cross-account-resource-access.html)
- [x] integrate 'account root' users into policy evaluation

## Bucket Ownership
- [x] `RGWBucketInfo::owner` uses rgw_owner variant which can either be rgw_user or rgw_account_id
- [x] buckets created by a user/role in an account are owned by the account
- [x] rgw_quota.cc accumulates/flushes account stats similar to user stats based on `RGWBucketInfo::owner`
- [x] ListBuckets can list from the owning account

## IAM APIs
- [x] CreateUser, GetUser, UpdateUser, DeleteUser, ListUsers apis (ex https://docs.aws.amazon.com/IAM/latest/APIReference/API_CreateUser.html)
- [x] AccessKey APIs
- [x] adapt existing Role APIs to account permission model
- [x] adapt existing UserPolicy APIs to account permission model
- [x] adapt existing Topic APIs to account permission model
- [x] adapt existing OpenIDConnectProvider APIs to account permission model
- [x] Group/GroupPolicy APIs

## Testing
- [x] add account root user to s3tests config for iam api tests
- [x] s3tests cases for new apis in test_iam.py (User apis started in https://github.com/ceph/s3-tests/pull/537)
- [x] s3tests for account ids in bucket policy
- [x] in teuthology, run s3tests in configurations where the 's3 main' and 's3 tenant' users belong to an account
- [x] in teuthology, create account-root users for s3test config [iam root] and [iam alt root] to run iam_account and cross_account tests
- [x] test user migration into an account and verify access to existing buckets/objects
- [x] test account stats and quota enforcement
- [x] multisite test cases for sync of account metadata and related indices
<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

anthonyeleven commented on doc/radosgw/account.rst:
```diff
@@ -0,0 +1,129 @@
+
+.. versionadded:: Squid
+
+Ceph Object Gateway supports user accounts as an optional feature to enable
```
The Ceph Object ...?  not sure what usage we want to converge on.

Maybe italicize _user accounts_?

Is this congruent with AWS S3, or an RGW-specific feature?

anthonyeleven commented on doc/radosgw/account.rst:
```diff
@@ -0,0 +1,129 @@
+Account Root User
+=================
+
+Like normal users and roles, accounts and account root users must be
```
Here you have "account root" but above "user accounts", should above be "root users" too?  "account" below as well, we should be consistent throughout the document.

cbodley replied:
thanks @anthonyeleven, i tried to clarify the language here. would you mind giving the docs another review?

anthonyeleven commented on doc/radosgw/account.rst:
```diff
@@ -0,0 +1,129 @@
+Like normal users and roles, accounts and account root users must be
+created by an administrator using ``radosgw-admin`` or the `Admin Ops API`_.
+
+The account root user has superuser permissions on all resources owned
```
s/superuser/admin/?

anthonyeleven commented on doc/radosgw/account.rst:
```diff
@@ -0,0 +1,129 @@
+The account root user has superuser permissions on all resources owned
+by the account. The root user's credentials (access and secret keys) can be
+used with the `Ceph Object Gateway IAM API`_ to create additional IAM users
+and roles for use with the `Ceph Object Gateway S3 API`_, as well as manage
```
as well as to

dang commented on src/rgw/rgw_sal.h:
```diff
@@ -461,11 +587,6 @@ class User {
 
     /** Clone a copy of this user.  Used when modification is necessary of the copy */
     virtual std::unique_ptr<User> clone() = 0;
-    /** List the buckets owned by a user */
```
Do we want the User to contain an rgw_owner now rather than an rgw_user?

cbodley replied:
no, each `sal::User` still corresponds to a unique `rgw_user`

the `rgw_owner` variant is for resources like buckets/objects that can either be owned by a user or an account

when a user is 'owned' by an account, we track that relationship with `RGWUserInfo::account_id`

dang reviewed (commented):
The SAL changes look fine in general.

pritha-srivastava commented on src/rgw/rgw_rest_role.cc:
```diff
@@ -551,96 +493,95 @@ void RGWModifyRoleTrustPolicy::execute(optional_yield y)
     const rgw::IAM::Policy p(
-      s->cct, s->user->get_tenant(), bl,
+      s->cct, s->user->get_tenant(), perm_policy,
       s->cct->_conf.get_val<bool>("rgw_policy_reject_invalid_principals"));
```
the policy verification that we do in rgw was tied to the `tenant` till now. with the `account` changes, shouldn't this be changed to `account`?

cbodley replied:
thanks @pritha-srivastava, i _just_ ran into problems here with the new s3test case `test_account_role_policy_allow_get_role()` for "test that identity policy can grant non-root users access to iam account apis"

because i passed in an empty tenant name there, the [// You can't specify resources for someone ELSE'S account](https://github.com/ceph/ceph/blob/ed736fab/src/rgw/rgw_iam_policy.cc#L628-L634) block causes us to parse `'Resource': '*'` as `Resource: [ arn:*:*:*::* ]` instead of `arn:*:*:*:*:*` so the comparison fails against role arns like `arn:aws:iam::RGW11111111111111111:role/s3-tests/s3-tests-MyRole`. i shared log snippet for this in https://gist.github.com/cbodley/96372128ea08e4c6c1f5479f887fc5a0

while i could pass in the account-id instead of the tenant to fix that case, cross-account access is a bigger issue. with accounts, the cross-account policy evaluation requires an identity policy to allow access to the resource, and a resource policy to allow access to the requesting identity. there it's crucial that identity policy be able to use a wildcard resource that covers other accounts. the resource policy part prevents that wildcard from granting access to things it shouldn't

for non-account users, we don't have those cross-account policy rules to save us, so we need that `// You can't specify resources for someone ELSE'S account` protection. but we should skip that part for account users

so i'm thinking about using `std::optional` for that tenant argument so that account users can pass in nullopt there. does that make sense @pritha-srivastava @adamemerson?

cbodley replied:
i pushed commit [rgw/iam: Policy takes optional tenant name](https://github.com/ceph/ceph/pull/54333/commits/f27bde2c8f2ec59d5fe541ff2da4df2ebacb21c0) for this and see the test passing now

pritha-srivastava commented on src/rgw/rgw_common.cc:
```diff
@@ -1133,6 +1142,87 @@ Effect eval_identity_or_session_policies(const DoutPrefixProvider* dpp,
+  ldpp_dout(dpp, 10) << __func__ << ": implicit deny from identity-based policy" << dendl;
+  return Effect::Pass;
+}
+
```
the iam policy verification changes wrt accounts looks fine to me.

cbodley commented:
sorry about all the extra squash commits; i'm still iterating on teuthology tests but don't want to rebase/squash while reviews are pending

adamemerson commented on src/rgw/rgw_iam_policy.h:
```diff
@@ -533,8 +590,8 @@ struct Policy {
-  Policy(CephContext* cct, const std::string& tenant,
-	 const bufferlist& text,
+  Policy(CephContext* cct, const std::string* tenant,
+	 std::string text,
```
I approve of this change.

adamemerson reviewed (approved):
This is a wonderful PR. Accounts give us something reified for multitenancy, and it adds much-needed features.

I also appreciate the factoring away from services.

pritha-srivastava commented on src/rgw/rgw_oidc_provider.cc:
```diff
@@ -1,182 +1,41 @@
+  encode_json("provider_url", provider_url, f);
+  encode_json("arn", arn, f);
+  encode_json("creation_date", creation_date, f);
+  encode_json("tenant", tenant, f);
```
do we need a check here that if the oidc provider belongs to a tenant, then display it, else don't

pritha-srivastava replied:
ok, got it - `tenant` represents both account and tenant

pritha-srivastava commented on src/rgw/rgw_auth.cc:
```diff
@@ -416,7 +510,10 @@ void rgw::auth::WebIdentityApplier::load_acct_info(const DoutPrefixProvider* dpp
+  ceph::real_time last_synced;
+  ceph::real_time last_updated;
+  int ret = driver->load_stats(dpp, null_yield, federated_user, stats,
+                               last_synced, last_updated);
```
@cbodley The part below where `create_account` needs to be looked at - a shadow user is created in rgw automatically derived from the `sub` field of the JWT, in case of AssumeRoleWithWebIdentity (for the first time), and stats are maintained for that user going forward. If resources are going to be tied to an account, then we might have to do away with these shadow users, if AssumeRoleWithWebIdentity is called for a role in an account.

cbodley replied:
for account roles, it's probably fine for WebIdentityApplier to stop writing shadow users to rados as long as load_acct_info() still initializes the RGWUserInfo for s->user which lots of ops still rely on

it also looks like there's more to do here to expose the role's account id to WebIdentityApplier so it can handle account ownership in get_aclowner() and is_owner_of()

for RemoteApplier, we don't define any mapping between remote users and rgw accounts, so any shadow users we create wouldn't belong to an account. i suppose an admin could come in and add the shadow users to an account if they wanted with `radosgw-admin user modify --account-id`. so if RemoteApplier::load_acct_info() finds a user with an account id, we should use that to handle account ownership in get_aclowner() and is_owner_of()

pritha-srivastava replied:
> for account roles, it's probably fine for WebIdentityApplier to stop writing shadow users to rados as long as load_acct_info() still initializes the RGWUserInfo for s->user which lots of ops still rely on
> 

if we stop creating shadow users for AssumeRoleWithWebIdentity, and we still need s->user, then what will the s->user point to, in other words, which user will we be referring to if we do away with the shadow users.

> it also looks like there's more to do here to expose the role's account id to WebIdentityApplier so it can handle account ownership in get_aclowner() and is_owner_of()
> 
> for RemoteApplier, we don't define any mapping between remote users and rgw accounts, so any shadow users we create wouldn't belong to an account. i suppose an admin could come in and add the shadow users to an account if they wanted with `radosgw-admin user modify --account-id`. so if RemoteApplier::load_acct_info() finds a user with an account id, we should use that to handle account ownership in get_aclowner() and is_owner_of()

cbodley replied:
WebIdentityApplier::create_account() is initializing some RGWUserInfo and writing it with store_user(). i'm guessing we could skip the store_user() part and expose that RGWUserInfo as-is

pritha-srivastava replied:
ok, yes, this will work as long as we do not need the bucket and user quota stats.

pritha-srivastava commented on src/rgw/rgw_rest_user_policy.cc:
```diff
@@ -38,377 +37,673 @@ void RGWRestUserPolicy::send_response()
+  s->info.args.remove("Action");
+  s->info.args.remove("Version");
+
+  int r = forward_iam_request_to_master(this, site, s->user->get_info(),
```
@cbodley : why do we explicitly need to send these requests for replication, won't the policies replicate as part of the user? or are these required for incremental sync?

pritha-srivastava replied:
What I am trying to ask is that, isnt this `forward_iam_request_to_master` taken care of in svc_user_rados.cc, where the user metadata is replicated in `PutOperation`.

cbodley replied:
when a secondary zone replicates user metadata from the master zone, it does call through RGWUserMetadataHandler::do_put() -> store_user_info() -> PutOperation

but if we write new metadata to a non-master zone, no other zones will try to replicate that. other zones only run metadata sync against the metadata master zone

forwarding metadata write requests to the master zone is an important part of our metadata consistency model

pritha-srivastava replied:
thank you for explaining this.

pritha-srivastava commented on src/rgw/rgw_rest_iam.h:
```diff
@@ -3,9 +3,92 @@
+  return r;
+}
+
+/// Perform an atomic read-modify-write operation on the given group metadata.
```
`role` metadata

cbodley commented on doc/radosgw/account.rst:
```diff
@@ -0,0 +1,232 @@
+:ref:`Multitenancy <rgw-multitenancy>` for details.
+
+A tenanted account can only contain users with the same tenant name.
+
```
i think namespace isolation is a useful feature. i'm fine with tenant if it does only that

cbodley commented on src/rgw/rgw_common.h:
```diff
@@ -742,6 +765,99 @@ struct RGWUserInfo
 WRITE_CLASS_ENCODER(RGWUserInfo)
 
+// user account metadata
+struct RGWAccountInfo {
```
i stuck with RGWAccountInfo and RGWGroupInfo to be consistent with other metadata types like RGWUserInfo and RGWBucketInfo

cbodley commented on doc/radosgw/account.rst:
```diff
@@ -0,0 +1,232 @@
+
+	radosgw-admin user modify --uid={userid} --account-id={accountid}
+
+.. note:: Ownership of all of the user's buckets will be transferred to
```
i don't see the potential for data loss here. we don't change any ACLs, so the user can probably still access everything even after being removed from the account. the buckets will just be accessible to other users in the account, and quota/usage stats are tracked by the account instead

cbodley commented on src/rgw/rgw_account.cc:
```diff
@@ -0,0 +1,391 @@
+  return id;
+}
+
+bool validate_id(std::string_view id, std::string* err_msg)
```
i wanted the string argument to be optional since we only need it for admin ops. other callers don't pay for the string allocation/formatting

cbodley commented:
tried running tempest tests in http://qa-proxy.ceph.com/teuthology/cbodley-2024-02-26_20:03:48-rgw:tempest-wip-rgw-account-v3-distro-default-smithi/7574168/teuthology.log and saw 3 tests fail with an unexpected BucketAlreadyExists error:

ContainerTest.test_create_container_overwrite
ContainerTest.test_create_container_with_remove_metadata_key
ContainerTest.test_create_container_with_remove_metadata_value

cbodley commented:
rebased/squashed

cbodley commented:
saw a `test_bucket_policy_put_obj_grant` failure in http://qa-proxy.ceph.com/teuthology/cbodley-2024-02-26_19:59:57-rgw:verify-wip-rgw-account-v3-distro-default-smithi/7574161/teuthology.log

verify_object_permission() wasn't applying the cross-account rules when the _object_ is owned by another account, it was only looking at the bucket owner

cbodley commented:
> tried running tempest tests in http://qa-proxy.ceph.com/teuthology/cbodley-2024-02-26_20:03:48-rgw:tempest-wip-rgw-account-v3-distro-default-smithi/7574168/teuthology.log and saw 3 tests fail with an unexpected BucketAlreadyExists error:
> 
> ContainerTest.test_create_container_overwrite
> ContainerTest.test_create_container_with_remove_metadata_key
> ContainerTest.test_create_container_with_remove_metadata_value

fixed. i had accidentally missed a negation for is_owner_of() in RGWCreateBucket

BBoozmen commented:
Running some tests again using the HEAD commit == b035dc8e6859594dc4486ba4adaf073909b0c1d0, I think there might be a regression caused in one of the recently added commits. With the following perm policy:

<pre>
...
        {
            "Effect": "Allow",
            "Action": ["s3:CreateBucket", "s3:DeleteBucket"],
            "Resource": f"arn:aws:s3:::u11b2",
        },
...
</pre>

The user (that belongs to the account who owns the role) assuming the role was able to create & delete the bucket but now delete fails (AccessDenied). In the log I see, it's referred as a cross-account access but it shouldn't be. The second line (inequality) makes it even more suspicious :) and the 3rd & 4th lines look contradictory.

<pre>
2024-02-28T21:20:06.144-0500 7f24ea073700  2 req 15383059350579586491 0.000000000s s3:delete_bucket verifying op permissions
2024-02-28T21:20:06.144-0500 7f24ea073700  4 req 15383059350579586491 0.000000000s s3:delete_bucket cross-account request for bucket owner RGW50444693535472184 != RGW50444693535472184
2024-02-28T21:20:06.144-0500 7f24ea073700 10 req 15383059350579586491 0.000000000s s3:delete_bucket evaluate_iam_policies: allowed by identity-based policy
2024-02-28T21:20:06.144-0500 7f24ea073700 10 req 15383059350579586491 0.000000000s s3:delete_bucket evaluate_iam_policies: implicit deny from identity-based
 policy
2024-02-28T21:20:06.144-0500 7f24ea073700  1 req 15383059350579586491 0.000000000s op->ERRORHANDLER: err_no=-13 new_err_no=-13
2024-02-28T21:20:06.144-0500 7f24ea073700 10 req 15383059350579586491 0.000000000s cache get: name=dev-zone-bcc-master.rgw.log++script.postrequest. : hit (n
egative entry)
2024-02-28T21:20:06.144-0500 7f24ea073700  2 req 15383059350579586491 0.000000000s s3:delete_bucket op status=0
2024-02-28T21:20:06.144-0500 7f24ea073700  2 req 15383059350579586491 0.000000000s s3:delete_bucket http status=403
2024-02-28T21:20:06.144-0500 7f24ea073700  1 ====== req done req=0x7f1fb2153490 op status=0 http_status=403 latency=0.000000000s ======
</pre>

pritha-srivastava commented:
@cbodley : My review of the PR is complete, I have no additional comments.

cbodley commented:
> Running some tests again using the HEAD commit == [b035dc8](https://github.com/ceph/ceph/commit/b035dc8e6859594dc4486ba4adaf073909b0c1d0), I think there might be a regression caused in one of the recently added commits.

thanks @BBoozmen, you're right. i had changed the owner comparisons to use is_owner_of(), but RoleApplier::is_owner_of() still had a `// TODO: handle account roles`. fixed

BBoozmen commented:
> some s3test cases in test_s3.py are failing because they don't get the expected DisplayName for acl grants; i've been using user names and role names there because the identities only know the account id. i'm planning to load the RGWAccountInfo and pass it into the identities instead so they can use the account's name for get_aclowner()

I think this might be the reason why scenario 1 from https://ceph-storage.slack.com/archives/C05LPHSKVPG/p1709151182645409 is still  observed. Just tested with the HEAD == 12b970a964518bb647ca04dbef2c6cea30fe9681, a user from another account can still access some other account's bucket. https://gist.github.com/BBoozmen/ac73c9b6914eaf19dd9bee242c74cf7b is the logs for the put request which should have failed.

cbodley commented:
> I think this might be the reason why scenario 1 from https://ceph-storage.slack.com/archives/C05LPHSKVPG/p1709151182645409 is still observed. Just tested with the HEAD == [12b970a](https://github.com/ceph/ceph/commit/12b970a964518bb647ca04dbef2c6cea30fe9681), a user from another account can still access some other account's bucket. https://gist.github.com/BBoozmen/ac73c9b6914eaf19dd9bee242c74cf7b is the logs for the put request which should have failed.

strange, thanks. i don't think that relates to the issue with acl DisplayNames, because permission checks don't look at the DisplayName

looking closer at RGWPutObj, i see that it's not calling `verify_object_permission()` where i implemented the cross-account rules. it's still calling `eval_identity_or_session_policies()` etc. manually

the same applies to all of RGWPutObj, RGWPostObj, RGWDeleteObj, RGWCopyObj, RGWInitMultipart, RGWCompleteMultipart, RGWAbortMultipart, RGWDeleteMultiObj, RGWBulkUploadOp

@pritha-srivastava do you remember why these can't use `verify_object_permission()` directly?

BBoozmen commented:
Some minor cosmetic issue...Even when an account deletion is successful, an error line is spit out:

```
$ sudo radosgw-admin account create --account-name=acc3
$ sudo radosgw-admin account rm --account-name=acc3
2024-02-29T16:30:50.969-0500 7fbd508a96c0  0 ERROR: obj.oid is empty
```

regardless of whether id or name is used in the command line.

cbodley commented:
> looking closer at RGWPutObj, i see that it's not calling verify_object_permission() where i implemented the cross-account rules. it's still calling eval_identity_or_session_policies() etc. manually
>
> the same applies to all of RGWPutObj, RGWPostObj, RGWDeleteObj, RGWCopyObj, RGWInitMultipart, RGWCompleteMultipart, RGWAbortMultipart, RGWDeleteMultiObj, RGWBulkUploadOp
>
> @pritha-srivastava do you remember why these can't use `verify_object_permission()` directly?

ok, the difference is that these ops need to evaluate iam policies according to the object's ARN, but fall back to the _bucket's_ acls on a pass. `verify_object_permission()` falls back to the object acl instead, which may not even exist for these ops

i'm thinking we should add a new `verify_bucket_permission()` overload that takes an ARN that be used instead of the bucket's. that way these ops can pass the object ARN and still take advantage of the cross-account policy evaluation

pritha-srivastava commented:
> verify_bucket_permission

ok, sounds fine.

pritha-srivastava commented:
> > looking closer at RGWPutObj, i see that it's not calling verify_object_permission() where i implemented the cross-account rules. it's still calling eval_identity_or_session_policies() etc. manually
> > the same applies to all of RGWPutObj, RGWPostObj, RGWDeleteObj, RGWCopyObj, RGWInitMultipart, RGWCompleteMultipart, RGWAbortMultipart, RGWDeleteMultiObj, RGWBulkUploadOp
> > @pritha-srivastava do you remember why these can't use `verify_object_permission()` directly?
> 
> ok, the difference is that these ops need to evaluate iam policies according to the object's ARN, but fall back to the _bucket's_ acls on a pass. `verify_object_permission()` falls back to the object acl instead, which may not even exist for these ops
> 
> i'm thinking we should add a new `verify_bucket_permission()` overload that takes an ARN that be used instead of the bucket's. that way these ops can pass the object ARN and still take advantage of the cross-account policy evaluation

ok, sounds fine.

cbodley commented on src/rgw/rgw_op.cc:
```diff
@@ -631,12 +689,11 @@ int rgw_build_bucket_policies(const DoutPrefixProvider *dpp, rgw::sal::Driver* d
+        load_iam_identity_policies(dpp, y, driver,
+                                   s->user->get_info(),
+                                   s->user->get_attrs(),
+                                   s->iam_identity_policies);
```
i'm seeing RGWPostObj auth failures with account users, because the authentication happens _after_ this call to `rgw_build_bucket_policies()`. so `s->user` at this point is the 'anonymous user', and we're unable to apply any of these identity policies

role policy gets added by `RoleApplier::modify_request_state()` so doesn't have this issue. i think the other Appliers should handle user policy the same way

BBoozmen replied:
Had a quick test of HEAD 1d033408c07388155c517eec957c6e591538e513. Things look good so far. I'll let our product team to resume their testing on this version.

cbodley replied:
> i'm seeing RGWPostObj auth failures with account users

fixed with commit "rgw/auth: pass user policies into identities"

cbodley commented:
s3tests are looking clean now. still debugging a couple of tempest test failures with swift FormPost

BBoozmen commented:
Not sure whether these requests belong to this PR or can be worked separately but ....

1. AFAICT there's no way to list users (inlcuding root user) of a given account: i.e., `user list --account-id=<>` doesn't list users of that account. This would be nice-to-have.

2. One other observation is - if an rgw user is linked to an account, it cannot be unlinked (e.g., using `user modify` with empty account-id) but would be linked to another account. Is there way to make an rgw user accountless again?

cbodley commented:
> Not sure whether these requests belong to this PR or can be worked separately but ....
> 
> 1. AFAICT there's no way to list users (inlcuding root user) of a given account: i.e., `user list --account-id=<>` doesn't list users of that account. This would be nice-to-have.

thanks, i agree. i had looked into that but it wasn't trivial. rgw_admin.cc shares the same logic for `metadata list`, `user list`, and `account list` in terms of `driver->meta_list_keys_next()`. i'll split out `user list` into its own command with a special case for `list_account_users()`

cbodley commented:
> 2. One other observation is - if an rgw user is linked to an account, it cannot be unlinked (e.g., using `user modify` with empty account-id) but would be linked to another account. Is there way to make an rgw user accountless again?

good question. i had considered disallowing the removal of account users once added, but it doesn't look i implemented that

the problem with removing users from an account is that they likely have identity policies attached, and once they're not in an account they don't respect the cross-account policy evaluation rules. that means their identity policies start granting them access to the resources of other users/accounts

moving users between accounts is also problematic in case they have group memberships (RGWUserInfo::group_ids would continue to reference the original account's groups)

i'm tempted to go with my original instinct to allow existing users to migrate into accounts, but not allow for their removal. i don't think there's a real benefit to moving one account's user into a different account, compared to using the iam apis to delete the one and create the other. account users don't really have any other state that's worth preserving, unlike non-account users which track a list of owned buckets and usage stats

what do you think?

BBoozmen commented:
> i'm tempted to go with my original instinct to allow existing users to migrate into accounts, but not allow for their removal. i don't think there's a real benefit to moving one account's user into a different account, compared to using the iam apis to delete the one and create the other. account users don't really have any other state that's worth preserving, unlike non-account users which track a list of owned buckets and usage stats

Agreed.

cbodley commented:
should be passing teuthology soon. next steps are to merge https://github.com/ceph/ceph/pull/55657 to main and rebase this on top (a rebase is already staged at https://github.com/cbodley/ceph/commits/wip-rgw-meta-topic-account/)

that rebase will let me finish up the final items:

- [ ] index account topics in the {account}.topics omap
- [ ] adapt existing Topic APIs to account permission model

i would like to merge this and get it backported to squid, then follow up with the final topic pieces separately. cc @yuvalif

cbodley commented:
the rgw/verify jobs are mostly green, but the [rgw/sts job](http://qa-proxy.ceph.com/teuthology/cbodley-2024-03-11_16:22:43-rgw-wip-rgw-account-v3-distro-default-smithi/7591485/teuthology.log) is failing with AccessDenied in lots of webidentity and session policy tests

BBoozmen commented:
Just tested the HEAD a8a7386c704eefccc6d4f6ca77fd5afd6a7a1f08 for quota management within an account. It looks like the the issue is still there. 

Test Case:

1. Create an account with quota enabled
<pre>
radosgw-admin account get --account-id=RGW06295800289909855
{
    "id": "RGW06295800289909855",
    "tenant": "",
    "name": "myacc",
    "email": "",
    "quota": {
        "enabled": true,
        "check_on_raw": false,
        "max_size": 1024,
        "max_size_kb": 1,
        "max_objects": 1
    },
    "max_users": 1,
    "max_roles": 1000,
    "max_groups": 1000,
    "max_buckets": 1000,
    "max_access_keys": 4
}
</pre>

2. Create a non-accounted user with no quotas and upload some data (e.g., 5 x 4KB objects)
3. Add this user as ROOT user to the account:
<pre>
radosgw-admin user modify --uid=myaccuser --account-root --account-id=RGW06295800289909855
</pre>

4. Although account's quota is now exceeded, adding user to the account succeeds - which is OK I think. This is not the issue.
5. However, subsequent PUTs succeed as well - added 9 more object afterwards totalling to 14 objs
<pre>
adosgw-admin -c ./ceph.conf account stats --account-id=RGW06295800289909855 --sync-stats
2024-03-11T15:46:07.903-0400 7f3001808700 -1 WARNING: all dangerous and experimental features are enabled.
{
    "stats": {
        "size": 57344,
        "size_actual": 57344,
        "size_utilized": 0,
        "size_kb": 56,
        "size_kb_actual": 56,
        "size_kb_utilized": 0,
        "num_objects": 14
    },
    "last_synced": "2024-03-11T19:46:08.000499Z",
    "last_updated": "2024-03-11T19:46:07.996392Z"
}
</pre>

Looking into the execution, I think we are still feeding the user-id (`myaccuser`) but not account id (`RGW06295800289909855`) to `get_owner_quota_info`.

<pre>
(gdb) list                                                                                                                                                                                                                             1432      }
1433                                                                                                                                                                                                                                   1434      RGWQuota user_quotas;                                                                                                                                                                                                        1435
1436      // consult the bucket owner's quota                                                                                                                                                                                          1437      int r = get_owner_quota_info(this, s->yield, driver,                                                                                                                                                                         1438                                   s->bucket_owner.id, user_quotas);                                                                                                                                                               1439      if (r < 0) {
1440        return r;                                                                                                                                                                                                                  1441      }                                                                                                                                                                                                                            (gdb) p s->bucket_owner.id                                                                                                                                                                                                             $12 = std::variant<rgw_user, rgw_account_id> [index 0] = {{tenant = "", ns = "", id = "myaccuser"}}
(gdb) p s->bucket_owner
$13 = {id = std::variant<rgw_user, rgw_account_id> [index 0] = {{tenant = "", ns = "", id = "myaccuser"}}, display_name = "myaccuser"}
</pre>

Owner (` s->owner`), though, is set correctly to the account id:

<pre>
(gdb) f 2
#2  RGWPutObj::init_processing (this=0x55b212d74a00, y=...) at ../src/rgw/rgw_op.cc:3898
3898      return RGWOp::init_processing(y);
(gdb) p s->bucket_owner
$15 = {id = std::variant<rgw_user, rgw_account_id> [index 0] = {{tenant = "", ns = "", id = "myaccuser"}}, display_name = "myaccuser"}
(gdb) p s->owner
$16 = {id = std::variant<rgw_user, rgw_account_id> [index 1] = {{<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >> = "RGW06295800289909855", <No data fields>}}, display_name = "myacc"}
</pre>

Since bucket is owned by the account-id
<pre>
radosgw-admin -c ./ceph.conf bucket stats --bucket=bucket1 | egrep "bucket|owner"
    "bucket": "bucket1",
    "owner": "RGW06295800289909855",
</pre>

I think how we set `s->bucket_owner` needs to be corrected I think.

BBoozmen commented:
Question regarding migration of non-accountant rgw users to an account. If the user has some buckets with bucket-policies, I believe the policies must be updated to include account-id in the ARN, right?

Just the way the user's all buckets' ownership is changed to be the account-id, do we have plans to traverse these buckets' policies to update the ARN? AFAIU from the changes we do not update the bucket policies when a user is migrated to an account.

cbodley commented:
> Question regarding migration of non-accountant rgw users to an account. If the user has some buckets with bucket-policies, I believe the policies must be updated to include account-id in the ARN, right?

@BBoozmen you mean the Principal ARNs, right? this is documented in https://ceph--54333.org.readthedocs.build/en/54333/radosgw/account/#principals

migrated users continue to match the tenant form of their ARN for backward compatibility here, so existing bucket policy should not need to change

cbodley commented:
> I think how we set `s->bucket_owner` needs to be corrected I think.

@BBoozmen thanks. i see that `bucket_owner` is coming from the bucket's RGW_ATTR_ACL, which Bucket::chown() didn't change during account migration. i added a change to reencode the new acl owner there

BBoozmen commented:
>  bucket_owner is coming from the bucket's RGW_ATTR_ACL, which Bucket::chown() didn't change during account migration. i added a change to reencode the new acl owner there

Tested the HEAD b7f11e0405e98f4f4f6d6f846bdf4b8f1f56f083 for account quota management...Things look good, thank you!

> you mean the Principal ARNs, right? this is documented in https://ceph--54333.org.readthedocs.build/en/54333/radosgw/account/#principals
>
> migrated users continue to match the tenant form of their ARN for backward compatibility here, so existing bucket policy should not need to change

Did a quick test. Yes, things MOSTLY seem to be working fine but I observed the following case where I think it may break the workflow for existing users:

Start with 2 accounts
- RGW24446460677198583 acc1 
- RGW35571596704969620 acc2
- and 2 NON-accounted users `acc1user2` and `acc2user2`. 

1. non-accounted rgw user (acc1user2) created a bucket: a1u2b1
2. gave itself and another non-accounted user (acc2user2) full access for a bucket
3. things work as expected...both users could access a1u2b1
4. then linked the acc1user2 to the account acc1
5. both acc1user2 and acc2user2 could still access a1u2b1
6. In addition acc1's root user can also access the bucket a1u2b1. Any other user couldn't access the bucket.

so far so good but after linking acc2user2 to acc2, it cannot access a1u2b1 anymore - which breaks the workflow. I'm not sure whether there's an easy solution for this problem.

cbodley commented:
> so far so good but after linking acc2user2 to acc2, it cannot access a1u2b1 anymore - which breaks the workflow. I'm not sure whether there's an easy solution for this problem.

@BBoozmen after migrating acc2user2 into the account, did you add any identity policies? the user should continue to match the bucket's acl grant, but would treat it as a cross-account request which also requires an Allow from identity policy

BBoozmen commented:
A minor issue I think...So, there's this `Description` field in the role:
```
role create --role-name=ReadAccess --path=/ ....
{
    "RoleId": "7e9121f5-3729-4d66-9d1b-9d8b3f22abb4",
    "RoleName": "ReadAccess",
    "Path": "/",
    "Arn": "arn:aws:iam:::role/ReadAccess",
    "CreateDate": "2024-03-13T16:49:28.798Z",
    "Description": "", <-------------------------- **Here**
    "MaxSessionDuration": 3600,
    "AssumeRolePolicyDocument": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Principal\":{\"AWS\":[\"arn:aws:iam::RGW33863762667553970:user/secondary\"]},\"Action\":[\"sts:AssumeRole\"]}]}",
    "AccountId": ""
}
```

However, radosgw-admin doesn't seem to provide `--description` flag to be used when creating or updating a role. So, this field always remains empty string.

cbodley commented:
i've opened https://github.com/cbodley/ceph/pull/7 to finish the last two todo items:

- [x] index account topics in the {account}.topics omap
- [x] adapt existing Topic APIs to account permission model

i kept that in a separate branch so it could be reviewed separately. depending on the timing, i may want to merge this accounts pr first and start running tests on squid. otherwise i can pull those commits into this pr once they've been reviewed

zdover23 commented on doc/radosgw/account.rst:
```diff
@@ -0,0 +1,243 @@
+
+Regardless of tenant, account IDs and email addresses must be globally unique.
+
+Account Management
```
@zdover23 of the future: Change all infinitive phrases in this section to imperative sentences. For example: s/To create an account:/To create an account, run a command of the following form:/

cbodley commented:
freshly rebased/squashed

bbjlomonaco commented on src/rgw/driver/rados/rgw_rest_user.cc:
```diff
@@ -237,6 +243,9 @@ void RGWOp_User_Create::execute(optional_yield y)
     op_state.set_system(system);
 
+  if (s->info.args.exists("account-root"))
+    op_state.set_account_root(account_root);
```
Documentation states that:
> Each account is managed by an account root user.

Should there be a check to see if there already is a root user on the account? Currently clients can create multiple root users for a single account, but documentation implies there should only be one root user.

cbodley replied:
yeah, i just didn't think it was worth the extra complexity to limit accounts to a single root user

the docs discuss what i expect to be the most common use case, where an admin creates an account and a single root user on behalf of an end-user

the "Migrate an existing User into an Account" section mentions that existing users could be migrated to account-root users, and i don't see a reason to prevent multiple such users from being added to the same account

what do you think?

bbjlomonaco replied:
For ceph I can't think of any concerns other then that its inconsistent with AWS accounts only having one root user per account. Other than that I see no downsides.

Only thing I would recommend is making the documentation a little more clear that an account could have one or more root users. When I first noticed the behavior I wasn't sure if it was intended. Thanks Casey!

bbjlomonaco commented on src/rgw/rgw_account.cc:
```diff
@@ -0,0 +1,529 @@
+  if (ret < 0) {
+    return ret;
+  }
+  if (!users.users.empty()) {
```
Should this be `!roles.roles.empty()`

cbodley commented:
> Conflicting files
src/rgw/CMakeLists.txt
src/test/rgw/test_d4n_filter.cc 

waiting to rebase until https://github.com/cbodley/ceph/pull/7 is ready

cbodley commented:
pulled in pubsub commits that were reviewed separately in https://github.com/cbodley/ceph/pull/7

cbodley commented:
sigh, need to work through some d4n-related breakage
```
[ 89%] Building CXX object src/rgw/CMakeFiles/rgw_common.dir/driver/d4n/rgw_sal_d4n.cc.o
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-3007-ge8600de4/rpm/el8/BUILD/ceph-19.0.0-3007-ge8600de4/src/rgw/driver/d4n/rgw_sal_d4n.cc: In member function ‘virtual int rgw::sal::D4NFilterObject::get_obj_attrs(optional_yield, const DoutPrefixProvider*, rgw_obj*)’:
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-3007-ge8600de4/rpm/el8/BUILD/ceph-19.0.0-3007-ge8600de4/src/rgw/driver/d4n/rgw_sal_d4n.cc:259:96: error: cannot convert ‘const rgw_owner’ {aka ‘const std::variant<rgw_user, rgw_account_id>’} to ‘const rgw_user&’
  259 |     std::unique_ptr<rgw::sal::User> user = this->driver->get_user(this->get_bucket()->get_owner());
      |                                                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
      |                                                                                                |
      |                                                                                                const rgw_owner {aka const std::variant<rgw_user, rgw_account_id>}
```

BBoozmen commented:
How is bucket quota managed for an account?

For a non-accounted rgw user, we can set user and bucket level quotas and both are visible in `user info`:

```
$ radosgw-admin user info --uid=ihummel | jq '.user_quota, .bucket_quota'
{
  "enabled": true,
  "check_on_raw": false,
  "max_size": 4028225472512,
  "max_size_kb": 3933813938,
  "max_objects": -1
}
{
  "enabled": true,
  "check_on_raw": false,
  "max_size": -1024,
  "max_size_kb": 0,
  "max_objects": 20000000
}
```

However, for an account, we can only set account level quota:

```
radosgw-admin account get --account-id=RGW12345678912345677 | jq '.quota'
{
        "enabled": false,
        "check_on_raw": false,
        "max_size": 2048,
        "max_size_kb": 2,
        "max_objects": -1
}
```

Shouldn't we also be able to set/view bucket quotas given an account?

BBoozmen commented:
> > Shouldn't we also be able to set/view bucket quotas given an account?
> 
> @BBoozmen you can set bucket quotas directly on the buckets, so this didn't seem useful to me. if you disagree, would you open a tracker issue with a feature request? i'm trying to get this pr merged for squid asap

Sounds good. We can discuss this as a separate item.

cbodley commented:
> The slave crashed or was restarted

cbodley commented:
jenkins test make check

cbodley commented:
the rgw/notifications tests are failing with `An error occurred (AuthorizationError) when calling the CreateTopic operation` after rebase over https://github.com/ceph/ceph/pull/55661, which changed `verify_user_permissions()` for subusers. i've opened https://github.com/ceph/ceph/pull/56838 to revert that change..

cbodley commented:
> the rgw/notifications tests are failing with An error occurred (AuthorizationError) when calling the CreateTopic operation after rebase over https://github.com/ceph/ceph/pull/55661, which changed verify_user_permissions() for subusers. i've opened https://github.com/ceph/ceph/pull/56838 to revert that change..

@yuvalif the new test_ps_s3_persistent_notification_http_account test case was failing after that. i had to add another commit to teach `wait_for_queue_to_drain()` to add the `--account-id` argument to `topic stats`

cbodley commented:
passed qa in https://pulpito.ceph.com/cbodley-2024-04-12_12:44:47-rgw-wip-rgw-account-v3-distro-default-smithi/ with rerun https://pulpito.ceph.com/cbodley-2024-04-12_16:16:29-rgw-wip-rgw-account-v3-distro-default-smithi/
